sbi/inference/base.py

```diff
@@ -0,0 +1,128 @@
+    """
+
+    def __init__(
+        self,
+        simulator: Callable,
+        prior: torch.distributions.Distribution,
+        true_observation: Tensor,
+        simulation_batch_size: int = 1,
+        device: Optional[torch.device] = None,
+        summary_writer: Optional[SummaryWriter] = None
+    ):
+
+        """
+        Args:
+            simulator: function that takes (vector) parameter and returns one  
+                (vector) result per parameter vector, with shape (num_parameters, result_dimension)
+            
+            prior: distribution-like object with `log_prob`and `sample` methods.
+            
+            true_observation: tensor containing the observation x0.
+            
+            simulation_batch_size: number of parameter vectors that the 
+                 simulator accepts and converts to data x at once. If -1, we simulate all parameter sets at the same time. If >= 1, the simulator has to process data of shape (sim_batch_size, parameter_dimension).
+            
+            summary_writer: an optional SummaryWriter to control log file 
+                location (default is <current working directory>/logs.)
+             
+            device: torch.device on which to compute (optional).
+            
+            mcmc_method: MCMC method to use for posterior sampling, one of 
+                ['slice', 'hmc', 'nuts'].
+        """
+
+        # check inputs: observation
+        self._true_observation = atleast_2d(true_observation)
+
+        # warn if prior is Uniform and dim_input > 1 (inferred by drawing once)
+        # XXX if prior is a torch Distribution, then it should support
+        # XXX event_shape, no need to sample
+        # XXX also, why use 'input' here? -> parms are input to the simulator
+        scalar_input = prior.sample().numel() == 1
+
+        if isinstance(prior, Uniform) and not scalar_input:
+            warnings.warn(
+                f"""The parameter dimension of the simulator we expect 
+                rom the provided prior's `event_shape` is >1,
+                and the prior is a PyTorch Uniform distribution. 
+                Beware that you are using a `batch_shape` >1 implicitly 
+                and `event_shape` 1, because PyTorch does not support 
+                multivariate Uniform. Please consider using a BoxUniform prior instead."""
+            )
+        # warning if observed data is multidimensional.
+        # XXX we want numel here?
+        if true_observation.squeeze().ndim > 1:
+            warnings.warn(
+                """`true_observation` has several dimensions. 
+                It could be e.g. a matrix (D=2) of observed data or a batch of
+                1D vector observations. SBI currently does not support batches."""
+            )
+
+        # XXX do simulator attributes get used outside of us accessing them from 
+        # XXX inference classes? In that case, we can store them in a dict
+        # XXX here, e.g. self._simspec()...
```

Comment:
I do not think that they are used, but I am not 100% sure either. Maybe @janfb knows?

Reply:
Since we may want to provide a user-friendly annotation system for simulators based on decorators, let us postpone handling of simulator meta-information to a later round of refactoring. @janfb, what is your take?

Reply:
They are used for logging. I used them to provide the same functionality as in the simulator classes.